core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -144,6 +149,23 @@ public final OSProcess get(int pid) {
      */
     public abstract void killAll(Map<String, String> modelEnvVars) throws InterruptedException;
 
+    private static List<String> classPath;
+
+    private final int softkillWaitSeconds = 2 * 60; // processes get at most 2 minutes to respond to SIGTERM
+
+    public static void init(ServletContextEvent event) {
+        // find "jenkins-core.jar" and store path to it in classPath, we'll need it to execute WinCtrlCSender
+        ServletContext servletContext = event.getServletContext();
+        final Set<String> resourcePaths = servletContext.getResourcePaths("/WEB-INF/lib/");
+        classPath = new ArrayList<String>();
+        for (final String resourcePath : resourcePaths) {
+            if (resourcePath.contains("jenkins-core") && resourcePath.endsWith(".jar")) {
+                String jarPath = servletContext.getRealPath(resourcePath);
```

Comment:
This does not smell right. Have you even tested this when using remote agents?

Reply:
I tested this with a local installation only. The issue to solve is to start a new process which attaches to the process we want to send Ctrl+C to. I'd be glad for suggestions since you have an elaborate nose.